### ISSUE TITLE
LDAP connection fails after upgrade to FOG 1.5.8

### ISSUE THREAD

**Issue by mrpouletcurry** [FOGProject/fogproject#385]
Problems after update fog project. The connection with a LDAP at Fog doesn't work.

The following parameters have been kept, after update :
- Server Address
- Search Basse DN
- Group Search DN
- Admin Group
- Mobile Group
- Bind DN
- Bind Password

**Comment by Sebastian-Roth** [comment]
@mrpouletcurry Which version of FOG did you use before the update and which version did you update to?

@eduardotseoane @george1422 You guys have any idea? I don't know of any changes in the LDAP code lately.

**Comment by mrpouletcurry** [comment]
Before update it was on 1.5.6 and update to 1.5.8.
We event move to develop branch 1.5.8.28

**Comment by Sebastian-Roth** [comment]
@eduardotseoane @George1422 Anyone an idea?

**Comment by mastacontrola** [comment]
I haven't a clue.

There should be indication of issue within the /var/log/<webpath>/<errorlog> file though as most things are now written there during the processing of LDAP requests.

Maybe that can give us all a better idea?

A few months ago we did modify the ports so we could configure 389 or 636. Maybe the port number for this is incorrect as of now? The reason for the switch was to have it be available to the admin to change as needed. This might have caused some issues?

To find the ports settings -> FOG Configuration -> LDAP_PORTS

In 1.6 the name of this is FOG_PLUGIN_LDAP_PORTS.

If you installed 1.6 verison then downgraded, maybe this caused some of the issue?

**Comment by Ssshafi** [comment]
Same here after update to 1.5.8.
Problem is the port number. But I can't find where I can correct this.

In database lsPort of my ldap is good (389), when I list in ldap plugin 389 is shown to, but if I edit my ldap server select box for ldap port is empty, I can't choose anything.

in /var/log/php-fpm/www-error.log I can see error : 
```
#7 /var/www/html/fog/management/index.php(31): ProcessLogin->processMainLogin()
#8 {main}
  thrown in /var/www/html/fog/lib/plugins/ldap/class/ldap.class.php on line 147
```
And in /var/log/httpd/sssl_error_log : 
` PHP Fatal error:  Uncaught Exception: Port is not valid ldap/ldaps port in /var/www/html/fog/lib/plugins/ldap/class/ldap.class.php:147\nStack trace:\n#0 /var/www/html/fog/lib/plugins/ldap/class/ldap.class.php(256): LDAP->_ldapUp()\n#1 /var/www/html/fog/lib/plugins/ldap/hooks/ldappluginhook.hook.php(126): LDAP->authLDAP('USER', 'PASS')\n#2 /var/www/html/fog/lib/fog/hookmanager.class.php(87): LDAPPluginHook->checkAddUser(Array)\n#3 /var/www/html/fog/lib/fog/user.class.php(144): HookManager->processEvent('USER_LOGGING_IN', Array)\n#4 /var/www/html/fog/lib/fog/user.class.php(226): User->passwordValidate('USER', 'PASS')\n#5 /var/www/html/fog/lib/fog/fogbase.class.php(2456): User->validatePw('USER', 'PASS')\n#6 /var/www/html/fog/lib/pages/processlogin.class.php(255): FOGBase::attemptLogin('USER', 'PASS')\n#7 /var/www/html/fog/management/index.php(31): ProcessLogin->processMainLogin()\n#8 {main}\n  thrown in /var/www/html/fog/lib/plugins/ldap/class/ldap.class.php on line 147, referer: https://fog/fog/management/index.php?node=home
`

Also I can't find : the ports settings ->  FOG Configuration -> LDAP_PORTS

**Comment by Ssshafi** [comment]
FOG_USER_FILTER and LDAP_PORTS Settings wasn't created in database by upgrade.
I remove LDAP Plugin, reinstall it and now it's OK.

**Comment by mastacontrola** [comment]
Sometimes this happes @Ssshafi 

While we do try to ensure the Core items have a proper upgrade path, plugins are not considered a part of the core. So when new features/improvements are made to a plugin specifically, there is a high likely-hood one would need to uninstall and reinstall the plugin.

For sanity sake, it would make much better sense to ensure you backup the plugin table or just create a full backup of the database if you'd like so that you at least have a reference to rebuild. While "breaking" changes within a plugin may happen, Generally they are minimal. For example, your LDAP_PORTS and USER_FILTER were added, and only impact what the port listing will be and what users you want filtered due to being "LDAP" users.  The breaking part was the LDAP_PORTS item though as we changed how the interface gets and uses the ports.

**Comment by mastacontrola** [comment]
@Sebastian-Roth I do see a way forward,

Right now, even though we allow the LDAP_PORTS to be user generated, there are checks to try to ensure the ports are either: 389 or 636. In case you want to use ldaps:// it will use port 636, and ldap will use 389.

I think we may need to add an almost breaking change to the ldap plugin once again, to add an "LDAPS" checkbox. Using this method, we can tell which path to use for ldap connection and incorporate the DSN style of making the connection:

e.g.:
```
$dsn = 'ldap' . ( $this->isLdaps ? 's' : '') . '://' . $this->get('address') . ':' . $this->get('port');
// stuff using this after
```

isLdaps would be a database field with ENUM('0', '1') (on or off) and a checkbox within the form field.

Let me know what you think. This would help out with people who use custom ports and make the plugin a little more flexible I think.

**Comment by mastacontrola** [comment]
I've added this thought to working-1.6 branch. You guys are all more than welcome to give this version or shot, or if @Sebastian-Roth thinks it's worth adding to dev-branch waiting for that. If you do decide to update, however, it would not be recommended to try to downgrade unless you have a snapshot from before the upgrade (any changes after snapshot would be lost) or a backup you don't mind reverting to before installing the downgraded version (again any changes to db after the backup would be lost).

To use this element in working-1.6 (and possibly for dev-branch/1.5.x if this is added) you would alter the table. In working-1.6 this would be:
```
ALTER TABLE LDAPServers ADD lsIsLDAPs ENUM('0','1') NOT NULL DEFAULT '0';
```

As I said, this is a sort of breaking change as it adds a new column to the LDAPServers table. Meaning, if you update your instance to a version that has this, without uninstalling/reinstalling the LDAP Plugin.

**Comment by Sebastian-Roth** [comment]
@mastacontrola Sorry for the delay, have been onto other things. Nice work you figured out what was causing the issue for @Ssshafi!

Yes seems like the LDAP ports 389/636 hard coded in PHP will cause problems sooner than later. It's great you have added a proper fix to `working-1.6`. I just added one comment to that commit.

Thinking about this for 1.5.x I decided we go the simple route here. See commit 05b22b7.
